File: 2041-lab-files/lab06/all_whales.pl
#!/usr/bin/perl -w

# read from stdin
my %whale_pods;
while ($line = <STDIN>) {
    chomp $line;
    ($num, $type) = split(/ /, $line, 2);
    # 2. convert to lowercase
    #   transliterate (same as y///), doesn't require suffix g
    $type =~ tr/[A-Z]/[a-z]/;
    # 3. deleting plurality (assume naive case is safe)
    $type =~ s/s$//;
    # 4. trim leading and trailing whitespace
    $type =~ s/^\s+//;
    $type =~ s/\s+$//;  # greedy by default
    # 4. reconvert whitespace inbetween
    $type =~ s/\s+/ /g;

    $whale_pods{$type}{"n_pods"} += 1;
    $whale_pods{$type}{"total"} += $num;
}

# 1. print in alphabetical order
foreach $key (sort keys %whale_pods) {
    $n_pods = $whale_pods{"$key"}{"n_pods"};
    $total = $whale_pods{"$key"}{"total"};
    print "$key observations: $n_pods pods, $total individuals\n";
}

# last 2 dp: means no hard coding

File: 2041-lab-files/lab06/all_whales.py
#!/usr/bin/env python3
import sys, re

all_whales = {}
for line in sys.stdin: # best method
    line = line.strip().lower()
    line = re.sub('s$', '', line)
    line = re.sub('\s+', ' ', line)
    n_whales, whale_type = line.split(' ', 1)
    # 1: means, num splits to be made
    n_whales = int(n_whales)
    if whale_type not in all_whales.keys():
        all_whales[whale_type] = [0,0]
    # a[0]: pods, a[1]: total
    all_whales[whale_type][0] += 1
    all_whales[whale_type][1] += n_whales

all_whale_keys = list(all_whales.keys())
all_whale_keys.sort()
for whale_type in all_whale_keys:
    print(whale_type, "observations:",
        all_whales[whale_type][0], "pods,",
        all_whales[whale_type][1], "individuals")
